tools/user_migration/migrate_user.py

```diff
@@ -0,0 +1,288 @@
+        description="move user from sql to spanner")
+    parser.add_argument(
+        '--dsns', default="move_dsns.lst",
+        help="file of new line separated DSNs")
+    parser.add_argument(
+        '--users', default="move_users.lst",
+        help="file of new line separated users to move")
+    parser.add_argument(
+        '--token_dsn',
+        help="DSN to the token server database (optional)"
+    )
+    parser.add_argument(
+        '--verbose',
+        action="store_true",
+        help="verbose logging"
+    )
+    return parser.parse_args()
+
+
+def conf_mysql(dsn):
+    """create a connection to the original storage system """
+    connection = connector.connect(
+        user=dsn.username,
+        password=dsn.password,
+        host=dsn.hostname,
+        port=dsn.port or 3306,
+        database=dsn.path[1:]
+    )
+    return connection
+
+
+def conf_spanner(dsn):
+    """create a connection to the new Spanner system"""
+    path = dsn.path.split("/")
+    instance_id = path[-3]
+    database_id = path[-1]
+    client = spanner.Client()
+    instance = client.instance(instance_id)
+    database = instance.database(database_id)
+    return database
+
+
+def conf_db(dsn):
+    """read the list of storage definitions from the file and create
+    a set of connetions.
+
+     """
+    if dsn.scheme == "mysql":
+        return conf_mysql(dsn)
+    if dsn.scheme == "spanner":
+        return conf_spanner(dsn)
+    raise RuntimeError("Unknown DNS type: {}".format(dsn.scheme))
+
+
+def update_token(databases, user):
+    """optionally update the TokenServer storage indicating the user
+    is now on Spanner
+
+    """
+    cursor = databases['token'].cursor()
+    cursor.execute(
+        """
+        UPDATE
+            users
+        SET
+            replaced_at = {timestamp},
+            nodeid = {nodeid}
+        WHERE
+           uid = {uid}
+        """.format(
+            timestamp=int(time.time() * 100),
+            nodeid=SPANNER_NODE_ID,
+            uid=user)
+        )
+
+
+# The following two functions are taken from browserid.utils
+def encode_bytes_b64(value):
+    return base64.urlsafe_b64encode(value).rstrip(b'=').decode('ascii')
+
+
+def format_key_id(keys_changed_at, key_hash):
+    return "{:013d}-{}".format(
+        keys_changed_at,
+        encode_bytes_b64(key_hash),
+    )
+
+
+def get_fxa_id(databases, user):
+    """generate the spanner user key values from the original storage
+    data.
+
+    """
+    sql = """
+        SELECT
+            email, generation, keys_changed_at, client_state
+        FROM users
+            WHERE uid = {uid}
+    """.format(uid=user)
+    cursor = databases['mysql'].cursor()
+    cursor.execute(sql)
+    (email, generation, keys_changed_at, client_state) = cursor.next()
+    fxa_uid = email.split('@')[0]
+    fxa_kid = format_key_id(
+        keys_changed_at or generation,
+        bytes.fromhex(client_state),
+    )
+    cursor.close()
```

Comment:
not worth changing now (unless we start using pypy for this) but I think these calls can be made via the higher level sqlalchemy connection object which should support the with statement (vs cursors)

Reply:
Probably. `with` should call close automatically. I was skipping that here to aid in debugging.